File: DATASTRUCTURES/src/main/python/mylib/datastructures/trees/AVL.py
from typing import Union
from datastructures.nodes.TNode import TNode


class AVL:
    def __init__(self, root: TNode = None):
        self.root = root

    def insert(self, val: int):
        new_node = TNode(val)
        if self.root is None:
            self.root = new_node
            return
        node, parent = self.root, None
        while node is not None:
            parent = node
            if val < node.data:
                node = node.left
            elif val > node.data:
                node = node.right
            else:
                return  # value already exists
        new_node.parent = parent
        if val < parent.data:
            parent.left = new_node
        else:
            parent.right = new_node
        self._rebalance(new_node)

    def insert_node(self, node: TNode):
        if self.root is None:
            self.root = node
            return
        parent = None
        current = self.root
        while current is not None:
            parent = current
            if node.data < current.data:
                current = current.left
            elif node.data > current.data:
                current = current.right
            else:
                return  # value already exists
        node.parent = parent
        if node.data < parent.data:
            parent.left = node
        else:
            parent.right = node
        self._rebalance(node)

    def delete(self, val: int):
        node = self.search(val)
        if node is None:
            print(f"Value {val} not found in tree")
            return
        if node.left is None or node.right is None:
            child = node.left or node.right
            if child is None:
                child = TNode()
            if node.parent is None:
                self.root = child
            elif node == node.parent.left:
                node.parent.left = child
            else:
                node.parent.right = child
            child.parent = node.parent
            node.parent = node.left = node.right = None
            self._rebalance(child)
        else:
            pred = node.left
            while pred.right is not None:
                pred = pred.right
            node.data = pred.data
            self.delete(pred.data)

    def search(self, val: int) -> Union[TNode, None]:
        node = self.root
        while node is not None:
            if node.data == val:
                return node
            elif val < node.data:
                node = node.left
            else:
                node = node.right
        return None

    def print_in_order(self):
        def _print_in_order(node):
            if node is not None:
                _print_in_order(node.left)
                print(node.data, end=' ')
                _print_in_order(node.right)

        _print_in_order(self.root)
        print()

File: src/main/java/mylib/datastructures/trees/AVL.java
package main.java.mylib.datastructures.trees;

import main.java.mylib.datastructures.nodes.TNode;

/**
 * AVL Tree class that extends BST
 */
public class AVL extends BST {
    private TNode root;

    public AVL() {
        this.root = null;
    }

    public AVL(int val) {
        this.root = new TNode(val,0,null,null,null);
    }

    public AVL(TNode obj) {
        if (obj == null) {
            this.root = null;
        } else {
            this.root = obj;
            balanceTree();
        }
    }

    public void setRoot(TNode root) {
        this.root = root;
        if (root != null) {
            balanceTree();
        }
    }

    public TNode getRoot() {
        return this.root;
    }

    @Override
    public void insert(int val) {
        super.insert(val);
        balanceTree();
    }

    @Override
    public void insert(TNode node) {
        super.insert(node);
        balanceTree();
    }

    @Override
    public void delete(int val) {
        TNode nodeToDelete = search(val);
        if (nodeToDelete == null) {
            System.out.println("Value " + val + " not found in the tree.");
            return;
        }
        super.delete(nodeToDelete.getData()); // needs to be fixed
        balanceTree();
    }

    private void balanceTree() {
        if (root != null) {
            root = balanceTree(root);
        }
    }

    private TNode balanceTree(TNode node) {
        if (node == null) {
            return null;
        }
        if (getHeight(node.getLeft()) - getHeight(node.getRight()) > 1) {
            if (getHeight(node.getLeft().getLeft()) >= getHeight(node.getLeft().getRight())) {
                node = rotateRight(node);
            } else {
                node.setLeft(rotateLeft(node.getLeft()));
                node = rotateRight(node);
            }
        } else if (getHeight(node.getRight()) - getHeight(node.getLeft()) > 1) {
            if (getHeight(node.getRight().getRight()) >= getHeight(node.getRight().getLeft())) {
                node = rotateLeft(node);
            } else {
                node.setRight(rotateRight(node.getRight()));
                node = rotateLeft(node);
            }
        }
        updateHeight(node);
        return node;
    }

    private TNode rotateLeft(TNode node) {
        TNode newRoot = node.getRight();
        node.setRight(newRoot.getLeft());
        newRoot.setLeft(node);
        updateHeight(node);
        updateHeight(newRoot);
        return newRoot;
    }

    private TNode rotateRight(TNode node) {
        TNode newRoot = node.getLeft();
        node.setLeft(newRoot.getRight());
        newRoot.setRight(node);
        updateHeight(node);
        updateHeight(newRoot);
        return newRoot;
    }

    private int getHeight(TNode node) {
        if (node == null) {
            return -1;
        } else {
            return node.getBalance();
        }
    }

    private void updateHeight(TNode node) {
        if (node != null) {
            node.setBalance(Math.max(getHeight(node.getLeft()), getHeight(node.getRight())) + 1);
        }
    }

}
